### ISSUE TITLE
Asyncio for external trigger function calls?

### ISSUE THREAD

**Issue by hjoliver** [cylc/cylc-flow#2917]
External trigger function calls are currently wrapped in a minimal main program and called as distinct concurrent subprocesses. (We can't run them in the main process because user-supplied functions may not be sufficiently light and quick, and Python's GIL makes threading more or less worthless).

Once we're at Python 3, ascyncio probably provides a better way to do this. (And would presumably also allow trigger functions to retain state between calls - which is currently not possible).

**Comment by matthewrmshin** [comment]
Ultimately, we'll need to refactor the main loop to work under asyncio. (Has there been some work on this already as part of #2825?)

**Comment by kinow** [comment]
I think @matthewrmshin is correct. If we use a web framework that uses asyncio, it would make sense to use its main loop too. If not as the replacement of the scheduler's main loop, then at least have a way to allow it to run in parallel (harder, but doable).

#2825 has a "working" version. So far I was the only one that tested it, and I used a simple suite I - for no special reason - always test cylc... the example tutorial cycling five. Can't tell what would happen putting that loop under stress...

**Comment by oliver-sanders** [comment]
Closing this as superseded by https://github.com/cylc/cylc-flow/issues/3497, though #3497 is kinda a duplicate of this...
